### PR TITLE
SNOW-1491175 Remove set_src_position, overhaul with_src_position for better performance

### PR DESCRIPTION
**Pull request by sfc-gh-heshah**
[SNOW-1491175](https://snowflakecomputing.atlassian.net/browse/SNOW-1491175)

Remove all uses of `set_src_position` and `get_first_non_snowpark_stack_frame`, update the necessary test cases, and use the `inspect` module with better practices including
- Deleting the retrieved frame to prevent reference cycles and memory leaks
- Avoid capturing code context for every file along the stack
- Skip two frames out of the Snowpark library code to retrieve relevant user code immediately when possible
- Add comments discussing functionality in detail for future improvements

[SNOW-1491175]: https://snowflakecomputing.atlassian.net/browse/SNOW-1491175?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/snowflake/snowpark/_internal/ast_utils.py [hunk 1]
```diff
@@ -333,84 +333,83 @@ def set_builtin_fn_alias(ast: proto.Expr, alias: str) -> None:
...
-
-
 assignment_re = re.compile(r"^\s*([a-zA-Z_]\w*)\s*=.*$", re.DOTALL)
 
-
 def with_src_position(
-    expr_ast: proto.Expr, assign: Optional[proto.Assign] = None
+    expr_ast: proto.Expr, 
+    assign: Optional[proto.Assign] = None,
+    max_stack_lookback: int = 3,
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/_internal/ast_utils.py:341]: This shouldn't be `max`. It should be the precise frame to look at.

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -581,8 +580,13 @@ def set_ast_ref(self, sp_dataframe_expr_builder: Any) -> None:
...
         Given a field builder expression of the AST type SpDataframeExpr, points the builder to reference this dataframe.
         """
+        if self._ast_id is None and FAIL_ON_MISSING_AST:
+            _logger.debug(self._explain_string())
+            raise NotImplementedError(
+                f"DataFrame with API usage {self._plan.api_calls} is missing complete AST logging."
+            )
         # TODO: remove the None guard below once we generate the correct AST.
-        if self._ast_id is not None:
+        elif self._ast_id is not None:
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/dataframe.py:589]: This should be just an `else`. You should fold the FAIL_ON_MISSING_AST_CHECK above into a nested if in the `then` branch of the `if` statement.

File: tests/ast/data/df_analytics_functions.test [hunk 3]
```diff
@@ -57,16 +57,16 @@ time_series_agg_res = df.analytics.time_series_agg(
 
 df = session.create_dataframe([["2023-01-01", 101, 200], ["2023-01-02", 101, 100], ["2023-01-03", 101, 300], ["2023-01-04", 102, 250]])
 
-res1 = df.to_df("ORDERDATE", "PRODUCTKEY", "SALESAMOUNT")
+df = df.to_df("ORDERDATE", "PRODUCTKEY", "SALESAMOUNT")
```
Comment (sfc-gh-azwiegincew) [inline tests/ast/data/df_analytics_functions.test:60]: Are you sure you're using the current unparser?

**Review by sfc-gh-azwiegincew — APPROVED** [review]

File: tests/ast/data/df_analytics_functions.test (hunk 3 above)
Comment (sfc-gh-heshah) [reply tests/ast/data/df_analytics_functions.test]: Oh, I think this PR might include all of the test updates from your addition of expression inlining to the Unparser now, is that okay?

Comment (sfc-gh-lspiegelberg) [reply tests/ast/data/df_analytics_functions.test:60]: updated with recent main, I think this should be not relevant anymore.

**Pull request merged**